core/git.go

```diff
@@ -283,22 +273,88 @@ func (repo *RemoteGitRepository) setup(ctx context.Context) (_ *gitutil.GitCLI,
 	if err != nil {
 		return nil, nil, err
 	}
+	var resolvPath string
 	if netConf != nil {
-		resolvPath, err := mountResolv(netConf)
+		resolvPath, err = mountResolv(netConf)
 		if err != nil {
 			return nil, nil, err
 		}
-		opts = append(opts, gitutil.WithExec(func(ctx context.Context, cmd *exec.Cmd) error {
-			return runWithStandardUmaskAndNetOverride(ctx, cmd, "", resolvPath)
-		}))
-		cleanups.Add("remove updated /etc/resolv", func() error {
-			return os.Remove(resolvPath)
-		})
+		cleanups.Add("remove updated /etc/resolv", func() error { return os.Remove(resolvPath) })
 	}
 
+	// Inject auth headers and DNS config into git commands.
+	// Auth: Uses GIT_CONFIG_* env vars to set http.extraheader for submodules.
+	// DNS: Binds custom resolv.conf in mount namespace for session consistency.
+	opts = append(opts, gitutil.WithExec(func(ctx context.Context, cmd *exec.Cmd) error {
+		if headerLine != "" {
+			base := repo.URL.Scheme + "://" + repo.URL.Host
+			entries := map[string]string{
+				"http." + base + "/.extraheader": headerLine,
+			}
+			cmd.Env = mergeGitConfigEnv(cmd.Env, entries)
+		}
+		return runWithStandardUmaskAndNetOverride(ctx, cmd, "", resolvPath)
+	}))
+
 	return gitutil.NewGitCLI(opts...), cleanups.Run, nil
 }
 
+// mergeGitConfigEnv merges key/value config pairs into the environment
+// using GIT_CONFIG_COUNT/KEY/VALUE. Preserves existing entries.
+// Keys are sorted for deterministic builds.
+func mergeGitConfigEnv(env []string, entries map[string]string) []string {
+	if len(entries) == 0 {
+		return env
+	}
+
+	const countPrefix = "GIT_CONFIG_COUNT="
+
+	next := 0
+	out := make([]string, 0, len(env)+len(entries)*2+1)
+	for _, e := range env {
+		if s, ok := strings.CutPrefix(e, countPrefix); ok {
+			if n, err := strconv.Atoi(s); err == nil && n > next {
+				next = n
+			}
+			continue // drop old COUNT; we’ll rewrite it
+		}
+		out = append(out, e)
+	}
+
+	// Deterministic: append in sorted key order.
+	keys := slices.Sorted(maps.Keys(entries))
+	for i, k := range keys {
+		v := entries[k]
+		idx := next + i
+		out = append(out,
+			"GIT_CONFIG_KEY_"+strconv.Itoa(idx)+"="+k,
+			"GIT_CONFIG_VALUE_"+strconv.Itoa(idx)+"="+v,
+		)
+	}
+
+	// Final COUNT == prior COUNT + number of new pairs.
+	out = append(out, countPrefix+strconv.Itoa(next+len(keys)))
+	return out
+}
+
+// buildBasicAuthHeader creates 'Authorization: Basic <base64(user:pass)>'.
+func buildBasicAuthHeader(host, explicitUser string, password []byte) string {
+	// caller-supplied username takes priority; otherwise pick a host-specific default
+	user := explicitUser
+	if user == "" {
+		switch host {
+		case "bitbucket.org":
+			// NOTE: bitbucket.org is picky, and needs *this* username
+			user = "x-token-auth"
+		default:
+			user = "x-access-token"
+		}
+	}
+	pw := strings.TrimSpace(string(password))
```

Comment:
Why space trimming here? If the user's specified a token that contains whitespace, we should use it.

Reply:
See https://github.com/dagger/dagger/pull/10836 - we should be doing secret whitespace trimming at source.

Reply:
I was not aware, the intent was the same, but implementation wrong, thanks 🙏